cosmology/GravityFieldSolver.hpp

```diff
@@ -0,0 +1,240 @@
+        // simply assumes them
+        if (this->getStype() == "CG") {
+            typedef ippl::BConds<Field<T, Dim>, Dim> bc_type;
+            bc_type allPeriodic;
+            for (unsigned int i = 0; i < 2 * Dim; ++i) {
+                allPeriodic[i] = std::make_shared<ippl::PeriodicFace<Field<T, Dim>>>(i);
+            }
+            phi_m->setFieldBC(allPeriodic);
+        }
+    }
+
+
+    /**
+     * @brief Run the solver based on the provided type
+     */
+    void runSolver() override {
+        if (this->getStype() == "CG") {
+            CGSolver_t<T, Dim>& solver = std::get<CGSolver_t<T, Dim>>(this->getSolver());
+            solver.solve();
+
+            if (ippl::Comm->rank() == 0) {
+                std::stringstream fname;
+                fname << "data/CG_";
+                fname << ippl::Comm->size();
+                fname << ".csv";
+
+                Inform log(NULL, fname.str().c_str(), Inform::APPEND);
+                int iterations = solver.getIterationCount();
+                // Assume the dummy solve is the first call
+                if (iterations == 0) {
+                    log << "residue,iterations" << endl;
+                }
+                // Don't print the dummy solve
+                if (iterations > 0) {
+                    log << solver.getResidue() << "," << iterations << endl;
+                }
+
+            }
+            ippl::Comm->barrier();
+        } else if (this->getStype() == "FFT") {
+            if constexpr (Dim == 2 || Dim == 3) {
+                std::get<FFTSolver_t<T, Dim>>(this->getSolver()).solve();
+            }
+        } else if (this->getStype() == "P3M") {
+            if constexpr (Dim == 3) {
+                std::get<P3MSolver_t<T, Dim>>(this->getSolver()).solve();
+            }
+        } else if (this->getStype() == "OPEN") {
+            if constexpr (Dim == 3) {
+                std::get<OpenSolver_t<T, Dim>>(this->getSolver()).solve();
+            }
+        } else {
+            throw std::runtime_error("Unknown solver type");
+        }
+    }
+
+
+    /**
+     * @brief Initialize the solver with parameters
+     * 
+     * @tparam Solver Solver type
+     * @param sp Parameter list for the solver
+     */
+    template <typename Solver>
+    void initSolverWithParams(const ippl::ParameterList& sp) {
+        this->getSolver().template emplace<Solver>();
+        Solver& solver = std::get<Solver>(this->getSolver());
+
+        solver.mergeParameters(sp);
+
+        solver.setRhs(*rho_m);
+
+        if constexpr (std::is_same_v<Solver, CGSolver_t<T, Dim>>) {
+            // The CG solver computes the potential directly and
+            // uses this to get the electric field
+            solver.setLhs(*phi_m);
+            solver.setGradient(*F_m);
+        } else {
+            // The periodic Poisson solver, Open boundaries solver,
+            // and the P3M solver compute the electric field directly
```

Comment:
comments here should change to reflect that this is the gravitational case instead of the electrostatics case (i.e. rho is the mass density, and we compute the gravitational force field and the gravitational potential, instead of tho being the charge density from which we get the electric field).